modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+    }
+}
+
+// Built-in kernels for OAK /////////////////////////////////////////////////////
+
+class GOAKBackendImpl final : public cv::gapi::GBackend::Priv {
+    virtual void unpackKernel(ade::Graph            &graph,
+                              const ade::NodeHandle &op_node,
+                              const cv::GKernelImpl &impl) override {
+        OAKGraph gm(graph);
+
+        const auto &kimpl  = cv::util::any_cast<cv::gimpl::GOAKKernel>(impl.opaque);
+        gm.metadata(op_node).set(cv::gimpl::OAKComponent{kimpl});
+    }
+
+    virtual EPtr compile(const ade::Graph &graph,
+                         const cv::GCompileArgs &args,
+                         const std::vector<ade::NodeHandle> &nodes,
+                         const std::vector<cv::gimpl::Data>& ins_data,
+                         const std::vector<cv::gimpl::Data>& outs_data) const override {
+        return EPtr{new cv::gimpl::GOAKExecutable(graph, args, nodes, ins_data, outs_data)};
+    }
+};
+
+cv::gapi::GBackend cv::gapi::oak::backend() {
+    static cv::gapi::GBackend this_backend(std::make_shared<GOAKBackendImpl>());
+    return this_backend;
+}
+
+namespace cv {
+namespace gimpl {
+namespace oak {
+
+// Kernels ///////////////////////////////////////////////////////////////
+
+template<class Impl, class K>
+class GOAKKernelImpl: public detail::OAKCallHelper<Impl, typename K::InArgs, typename K::OutArgs>
+                    , public cv::detail::KernelTag {
+    using P = detail::OAKCallHelper<Impl, typename K::InArgs, typename K::OutArgs>;
+public:
+    using API = K;
+    static cv::gapi::GBackend   backend() { return cv::gapi::oak::backend();  }
+    static GOAKKernel kernel()  { return GOAKKernel(&P::construct); }
+};
+
+#define GAPI_OAK_KERNEL(Name, API) \
+    struct Name: public cv::gimpl::oak::GOAKKernelImpl<Name, API>
+
+namespace {
+GAPI_OAK_KERNEL(GOAKEncFrame, cv::gapi::oak::GEncFrame) {
+    static void put(const std::unique_ptr<dai::Pipeline>& pipeline,
+                    const std::tuple<int, int>&,
+                    std::shared_ptr<dai::Node>& node,
+                    std::vector<std::pair<std::string, dai::Buffer>>&,
+                    GOAKContext::InputPtr* in,
+                    const cv::gapi::oak::EncoderConfig& cfg,
+                    GOAKContext::OutputPtr* out) {
+        auto videoEnc = pipeline->create<dai::node::VideoEncoder>();
+
+        // FIXME: convert all the parameters to dai
+        videoEnc->setDefaultProfilePreset(cfg.width, cfg.height,
+                                          cfg.frameRate,
+                                          dai::VideoEncoderProperties::Profile::H265_MAIN);
+        node = videoEnc;
+        *in = &(videoEnc->input);
+        *out = &(videoEnc->bitstream);
+    }
+};
+
+GAPI_OAK_KERNEL(GOAKSobelXY, cv::gapi::oak::GSobelXY) {
+    static void put(const std::unique_ptr<dai::Pipeline>& pipeline,
+                    const std::tuple<int, int>& camera_wh,
```

Comment:
why do you need this in every kernel?

Reply:
Some of the DAI operations need to set max resolution size